### PR TITLE
Bugfix: ensure reentryability for Node::setWorld* functions

### PR DESCRIPTION
**Pull request by PatriceJiang**
### Changelog

* Avoid errors when calling the setWorld* interface consecutively with different parameters.
* bugfix: Node::setWorldScale C++ impl sync with ts version

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by lealzhan — APPROVED** [review]

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

File: native/cocos/core/scene-graph/Node.cpp [hunk 1]
```diff
@@ -570,13 +569,13 @@ void Node::setWorldRotation(float x, float y, float z, float w) {
 
     _eulerDirty = true;
 
+    notifyLocalRotationUpdated();
```
Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp:572]: Why invoke it here?

File: native/cocos/core/scene-graph/Node.cpp [hunk 2]
```diff
@@ -600,8 +600,8 @@ void Node::setWorldScale(float x, float y, float z) {
         // convert to Matrix 3 x 3
         Mat3::fromMat4(tmpLocalTransform, &localRS);
         Mat3::fromQuat(_localRotation.getConjugated(), &localRotInv);
-        // remove rotation part of the local matrix 
-        Mat3::multiply(localRS, localRotInv, &localRS);
+        // remove rotation part of the local matrix
+        Mat3::multiply(localRotInv, localRS, &localRS);
```
Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp:603]: Why change it?

File: native/cocos/core/scene-graph/Node.cpp (hunk 1 above)
Comment (PatriceJiang) [reply native/cocos/core/scene-graph/Node.cpp:572]: Ensure js callback triggered by emit can see the value of new rotation. `setWorldPostion` & `setWorldScale` do act this too.

File: native/cocos/core/scene-graph/Node.cpp (hunk 2 above)
Comment (PatriceJiang) [reply native/cocos/core/scene-graph/Node.cpp:603]: Sync with the logic in ts

**Review by minggo — APPROVED** [review]

File: cocos/scene-graph/node.ts [hunk 3]
```diff
@@ -2353,17 +2349,19 @@ export class Node extends CCObject implements ISchedulable, CustomSerializable {
      * @param y Y axis scale
      * @param z Z axis scale
      */
-    public setWorldScale (x: number, y: number, z: number): void;
+    public setWorldScale(x: number, y: number, z: number): void;
 
     public setWorldScale (val: Vec3 | number, y?: number, z?: number) {
+        const parent = this._parent;
+        if (parent) {
+            this.updateWorldTransform();
```
Comment (dumganhar) [inline cocos/scene-graph/node.ts:2357]: Why call this.updateWorldTransform rather than parent.updateWorldTransform?

**Comment by dumganhar** [comment]
> Avoid errors when calling the setWorld* interface consecutively with different parameters.

What kinds of errors?

**Pull request merged**